pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,475 @@
+	nviAttrImageSize
+	nviAttrTextureNormalized
+	nviAttrSamplerInit
+	nviAttrParamCbank
+	nviAttrSmemParamOffsets
+	nviAttrCbankParamOffsets
+	nviAttrSyncStack
+	nviAttrTexidSampidMap
+	nviAttrExterns
+	nviAttrReqntid
+	nviAttrFrameSize
+	nviAttrMinStackSize
+	nviAttrSamplerForceUnnormalized
+	nviAttrBindlessImageOffsets
+	nviAttrBindlessTextureBank
+	nviAttrBindlessSurfaceBank
+	nviAttrKparamInfo
+	nviAttrSmemParamSize
+	nviAttrCbankParamSize
+	nviAttrQueryNumattrib
+	nviAttrMaxregCount
+	nviAttrExitInstrOffsets
+	nviAttrS2rctaidInstrOffsets
+	nviAttrCrsStackSize
+	nviAttrNeedCnpWrapper
+	nviAttrNeedCnpPatch
+	nviAttrExplicitCaching
+	nviAttrIstypepUsed
+	nviAttrMaxStackSize
+	nviAttrSuqUsed
+	nviAttrLdCachemodInstrOffsets
+	nviAttrLoadCacheRequest
+	nviAttrAtomSysInstrOffsets
+	nviAttrCoopGroupInstrOffsets
+	nviAttrCoopGroupMaxRegids
+	nviAttrSw1850030War
+	nviAttrWmmaUsed
+	nviAttrHasPreV10Object
+	nviAttrAtomf16EmulInstrOffsets
+	nviAttrAtom16EmulInstrRegMap
+	nviAttrRegcount
+	nviAttrSw2393858War
+	nviAttrIntWarpWideInstrOffsets
+	nviAttrSharedScratch
+	nviAttrStatistics
+
+	// New between cuda 10.2 and 11.6
+	nviAttrIndirectBranchTargets
+	nviAttrSw2861232War
+	nviAttrSwWar
+	nviAttrCudaApiVersion
+	nviAttrNumMbarriers
+	nviAttrMbarrierInstrOffsets
+	nviAttrCoroutineResumeIdOffsets
+	nviAttrSamRegionStackSize
+	nviAttrPerRegTargetPerfStats
+
+	// New between cuda 11.6 and 11.8
+	nviAttrCtaPerCluster
+	nviAttrExplicitCluster
+	nviAttrMaxClusterRank
+	nviAttrInstrRegMap
+)
+
+// nvInfoItem is the in-file representation of an item  header in the .nv.info section. The value follows
+// according to the format
+type nvInfoItem struct {
+	Format nvInfoFormat
+	Attr   nvInfoAttr
+}
+
+// cubinParser is a helper struct to parse the cubin ELF sections
+type cubinParser struct {
+	kernels            map[string]*CubinKernel
+	sectPrefixToParser map[string]func(*elf.Section, *bytes.Reader, string) error
+}
+
+func newCubinParser() cubinParser {
+	cp := cubinParser{
+		kernels:            make(map[string]*CubinKernel),
+		sectPrefixToParser: make(map[string]func(*elf.Section, *bytes.Reader, string) error),
+	}
+
+	cp.sectPrefixToParser[".nv.info"] = cp.parseNvInfoSection
+	cp.sectPrefixToParser[".text"] = cp.parseTextSection
+	cp.sectPrefixToParser[".nv.shared"] = cp.parseSharedMemSection
+	cp.sectPrefixToParser[".nv.constant"] = cp.parseConstantMemSection
+
+	return cp
+}
+
+func (cp *cubinParser) getKernel(name string) *CubinKernel {
+	if _, ok := cp.kernels[name]; !ok {
+		cp.kernels[name] = &CubinKernel{
+			Name: name,
+		}
+	}
+	return cp.kernels[name]
+}
+
+func (cp *cubinParser) parseCubinElf(data []byte) error {
+	// Hacks to be able to parse the ELF: the ELF version is not supported by the Go ELF parser, so we need to
+	// trick it into thinking it's the old version. Check for boundaries first
+	if len(data) < 21 {
+		return fmt.Errorf("invalid cubin data, too short")
+	}
+	data[20] = 1
+
+	cubinElf, err := elf.NewFile(bytes.NewReader(data))
```

Comment:
can this be memory exhaustive? what if the payload is a very large one?

Reply:
Cubin payloads should be manageable. I tried to use the readers, but there are two problems with that: one, that we need to modify the ELF header so that golang's library recognizes it, and also that we'd still need to load in-memory if we end up with compressed data.

Reply:
resolving this, as i couldn't find a good way to use the debug/elf pkg, it requires ReaderAt, and keeping the original stream ends up in ReadSeeker (for uncompressed section) or Reader (for compressed section - `lz4.NewCompressedReader`)

in case we need to optimize this pkg, we might have to ditch the debug/elf and implement our own custom parser. For now it would be an overkill